#### 目录介绍
- 01.详细崩溃日志信息
- 02.查看崩溃类信息
- 03.关于使用场景分析
- 04.项目中该异常分析
- 05.追溯导航源码链路
- 06.bug解决方案实践
- 07.其他延伸分析说明



### 01.详细崩溃日志信息
- 详细崩溃日志信息
    - 由安全管理器抛出，以指示安全违规。大部分集中在Android10和Android11系统，且在后台崩溃占比很大。
    ```
    SecurityException: Not allowed to delete channel didi_delivery_service with a foreground service
    ```
- 影响面分析
    - crash超过5千次，最多的崩溃bug。从48到62一直存在，前台崩溃34%，后台崩溃65%。
- 详细崩溃堆栈日志。该类型bug崩溃超过：5千次
    ```
    java.lang.SecurityException: Not allowed to delete channel didi_delivery_service with a foreground service
        at android.os.Parcel.createExceptionOrNull(Parcel.java:2376)
        at android.os.Parcel.createException(Parcel.java:2360)
        at android.os.Parcel.readException(Parcel.java:2343)
        at android.os.Parcel.readException(Parcel.java:2285)
        at android.app.INotificationManager$Stub$Proxy.deleteNotificationChannel(INotificationManager.java:4022)
        at android.app.NotificationManager.deleteNotificationChannel(NotificationManager.java:909)
        at com.google.android.libraries.navigation.internal.jo.b.a(PG:77)
        at com.google.android.libraries.navigation.internal.jo.b.a(PG:26)
        at com.google.android.libraries.navigation.internal.hk.a$c.a(PG:11)
        at com.google.android.libraries.navigation.internal.hk.a.a(PG:101)
        at com.google.android.libraries.navigation.internal.hk.a.a(PG:97)
        at com.google.android.libraries.navigation.internal.hk.g.a(PG:75)
        at com.google.android.libraries.navigation.internal.hk.n.b(PG:37)
        at com.google.android.libraries.navigation.internal.hk.n.a(PG:8)
        at com.google.android.libraries.navigation.internal.hk.s.a(PG:13)
        at com.google.android.libraries.navigation.internal.lm.m.b(PG:18)
        at com.google.android.libraries.navigation.internal.lm.h.b(PG:111)
        at com.google.android.libraries.navigation.internal.lm.n.run(PG:10)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:236)
        at android.app.ActivityThread.main(ActivityThread.java:8051)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:620)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1011)
    Caused by: android.os.RemoteException: Remote stack trace:
        at com.android.server.notification.NotificationManagerService$10.enforceDeletingChannelHasNoFgService(NotificationManagerService.java:3436)
        at com.android.server.notification.NotificationManagerService$10.deleteNotificationChannel(NotificationManagerService.java:3449)
        at android.app.INotificationManager$Stub.onTransact(INotificationManager.java:1729)
        at android.os.Binder.execTransactInternal(Binder.java:1159)
        at android.os.Binder.execTransact(Binder.java:1123)
    ```
- 还有一种日志，详细崩溃堆栈日志。该类型bug崩溃超过：2百次
    ```
    android.app.RemoteServiceException: Bad notification for startForeground: java.lang.RuntimeException: invalid channel for service notification: Notification(channel=didi_delivery_service pri=0 contentView=null vibrate=null sound=null tick defaults=0x0 flags=0x62 color=0x00000000 vis=PRIVATE)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1774)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6543)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:440)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:810)
    ```



### 02.查看崩溃类信息
- SecurityException
    ```
    public class SecurityException extends RuntimeException {
        private static final long serialVersionUID = 6878364983674394167L;
        public SecurityException() {
            super();
        }
        public SecurityException(String s) {
            super(s);
        }
        public SecurityException(String message, Throwable cause) {
            super(message, cause);
        }
        public SecurityException(Throwable cause) {
            super(cause);
        }
    }
    ```


### 03.关于使用场景分析
#### 3.1 项目使用场景分析
- 根据日志推理出项目中使用场景
    - 初始化：RiderApplicationLifeCycle-->keepAliveServiceSetUp()--->startRiderService()
    - 开启Service：RiderService-->startRiderService()--->startForegroundService()。【该处开启：RiderForegroundService】
    - 关闭Service：RiderService-->stopRiderService()--->stopService()。【该处开启：RiderForegroundService】
- RiderService分析
    - 其实就是一个IntentService，相当于执行完任务就自动结束和销毁。
- RiderForegroundService
    - 是一个Service，可以看到最终开启通知是调用Service中的startForeground方法
    - 那么思考一下，如何关闭该Service？就会关闭通知栏呢，这个是怎么做到的？
- 思考：RiderForegroundService如何关闭通知栏
    - 关于通知栏常见方式：调用cancel方法，点击销毁通知栏等。那么通过测试可知，当stop该service，则通知栏会销毁？


#### 3.2 为何使用Service打开通知栏
- 为了保活操作，开启前台Service打开通知栏思考？
    - Service几乎都是在后台运行的，所以Service系统优先级还是比较低的，当系统出现内存不足情况时，就有可能回收掉正在后台运行的Service。
    - 如果希望Service可以一直保持运行状态，而不会由于系统内存不足的原因导致被回收，那么就要提高Service的优先级，而提高优先级的方法有多种，其中一种就是考虑使用前台Service。
- 如何把Service设置成前台Service？
    - 很简单，使用startForeground即可；要取消前台，使用stopForeground即可。
- 测试，在Service的onDestroy销毁通知栏
    - 测试案例1：销毁时调用stopForeground(true)，发现要过几秒才会移除通知栏，这是为什么呢？
    - 测试案例2：销毁时什么都不操作，发现也可以移除通知栏，但是并没有调用stopForeground，这是怎么实现的呢？



#### 3.3 Service被杀如何去掉通知栏
- 第一种情况Service被stop
    - 待分析
- 第二种情况Service被kill
    - 


### 04.项目中该异常分析
#### 4.1 项目中异常分析
- 崩溃重要日志
    - java.lang.SecurityException：Not allowed to delete channel didi_delivery_service with a foreground service
    - Caused by: android.os.RemoteException: Remote stack trace:
- 根据android.os.RemoteException: Remote stack trace:找到抛出错误源码
    - 具体找到：Parcel 类中的 readException方法
    ```
    public final void readException(int code, String msg) {
        String remoteStackTrace = null;
        final int remoteStackPayloadSize = readInt();
        if (remoteStackPayloadSize > 0) {
            remoteStackTrace = readString();
        }
        Exception e = createException(code, msg);
        // Attach remote stack trace if availalble
        if (remoteStackTrace != null) {
            RemoteException cause = new RemoteException(
                    "Remote stack trace:\n" + remoteStackTrace, null, false, false);
            try {
                Throwable rootCause = ExceptionUtils.getRootCause(e);
                if (rootCause != null) {
                    rootCause.initCause(cause);
                }
            } catch (RuntimeException ex) {
                Log.e(TAG, "Cannot set cause " + cause + " for " + e, ex);
            }
        }
        SneakyThrow.sneakyThrow(e);
    }
    ```
- 根据java.lang.SecurityException：Not allowed to delete channel didi_delivery_service with a foreground service找到抛出异常
    - 只能根据崩溃堆栈日志找到：createException(Parcel.java:2369)来推理崩溃的错误源码。
    ```
    private Exception createException(int code, String msg) {
        switch (code) {
            case EX_SECURITY:
                return new SecurityException(msg);
        }
    }
    ```
- 寻找问题突破口
    - 从上面日志看的出来，最后都是执行到 Parcel 类中来，才抛出的异常。
    - 寻找问题重点突破口放到：deleteNotificationChannel
- 那些地方调用了：NotificationManager类中方法deleteNotificationChannel
    - SystemNotificationChannels--->removeDeprecated()--->deleteNotificationChannel(DEVICE_ADMIN_DEPRECATED)
    - NotificationChannelController--->createAll()--->deleteNotificationChannel(CHANNEL_ID_MOBILE_DATA_ALERT_DEPRECATED)
    - AlertWindowNotification---->onCancelNotification()---->deleteNotificationChannel(mNotificationTag)
    - NotificationChannels---->createAll()---->deleteNotificationChannel(SCREENSHOTS_LEGACY)
- 谷歌开启导航后通知栏
    - 谷歌开启导航后，这个时候会创建一个通知栏用来显示导航信息。那么谷歌究竟是怎么常见通知栏呢？
- 提出假设和猜想
    - 猜想1：咱app开启了一个前台服务用于保活，谷歌开始导航时也开启了一个前台服务通知栏提示导航信息，会相互影响吗？
    - 猜想2：谷歌导航sdk内部调用deleteNotificationChannel来删除渠道，导致保活这个通知被干掉后产生崩溃？




### 05.追溯导航源码链路
- 先看一下：com.google.android.libraries.navigation.internal.jo.b.a
    - 谷歌导航是一个黑盒，这个是结合谷歌1.18的代码来查看的。重点看：deleteNotificationChannel
    ```
    private static void a(List<v> var0, NotificationManager var1) {
        ex var2 = a(var0);
        lz var4 = (lz)ku.b(a(var1), var2).iterator();
    
        while(var4.hasNext()) {
            String var5 = (String)var4.next();
            var1.deleteNotificationChannel(var5);
        }
    }
    ```
- 首先看一下NotificationManager类的deleteNotificationChannel方法
    ```
    public void deleteNotificationChannel(String channelId) {
        INotificationManager service = getService();
        try {
            service.deleteNotificationChannel(mContext.getPackageName(), channelId);
        } catch (RemoteException e) {
            throw e.rethrowFromSystemServer();
        }
    }
    ```
- 接着网上追溯，INotificationManager是一个接口，具体看NotificationManagerService
    - 大概意思是：通过binder获取uid，
    ```
    @Override
    public void deleteNotificationChannel(String pkg, String channelId) {
        checkCallerIsSystemOrSameApp(pkg);
        final int callingUid = Binder.getCallingUid();
        if (NotificationChannel.DEFAULT_CHANNEL_ID.equals(channelId)) {
            throw new IllegalArgumentException("Cannot delete default channel");
        }
        cancelAllNotificationsInt(MY_UID, MY_PID, pkg, channelId, 0, 0, true,
                UserHandle.getUserId(callingUid), REASON_CHANNEL_BANNED, null);
        mRankingHelper.deleteNotificationChannel(pkg, callingUid, channelId);
        mListeners.notifyNotificationChannelChanged(pkg,
                UserHandle.getUserHandleForUid(callingUid),
                mRankingHelper.getNotificationChannel(pkg, callingUid, channelId, true),
                NOTIFICATION_CHANNEL_OR_GROUP_DELETED);
        savePolicyFile();
    }
    ```
- 接着追溯到Parcel的readException。可以发现最终是执行到createException创建了一个SecurityException异常。
    ```
    public final void readException() {
        int code = readExceptionCode();
        if (code != 0) {
            String msg = readString();
            readException(code, msg);
        }
    }
    
    public final void readException(int code, String msg) {
        String remoteStackTrace = null;
        final int remoteStackPayloadSize = readInt();
        if (remoteStackPayloadSize > 0) {
            remoteStackTrace = readString();
        }
        Exception e = createException(code, msg);
    }
    
    
    private Exception createException(int code, String msg) {
        switch (code) {
            case EX_SECURITY:
                return new SecurityException(msg);
        }
        return new RuntimeException("Unknown exception code: " + code
                + " msg " + msg);
    }
    ```




### 06.bug解决方案实践
- 解决方案




### 07.其他延伸分析说明





